File: src/cursor_orch/api/cursor_client.py
from __future__ import annotations

import logging
import random
import time
from dataclasses import dataclass

import requests

logger = logging.getLogger(__name__)

BASE_URL = "https://api.cursor.com"

INITIAL_DELAY = 5.0
BACKOFF_MULTIPLIER = 2.0
MAX_DELAY = 60.0
MAX_RETRIES_429 = 5
MAX_RETRIES_TRANSIENT = 3
JITTER_FACTOR = 0.2
TRANSIENT_CODES = (502, 503, 504)


class AgentAPIError(Exception):
    def __init__(self, message: str, status_code: int = 0):
        super().__init__(message)
        self.message = message
        self.status_code = status_code


class AgentNotFound(AgentAPIError):
    def __init__(self, agent_id: str = ""):
        super().__init__(f"Agent not found: {agent_id}", 404)


class RateLimitError(AgentAPIError):
    def __init__(self, message: str = "Rate limit exceeded", retry_after: float = 0):
        super().__init__(message, 429)
        self.retry_after = retry_after


@dataclass
class AgentInfo:
    id: str
    name: str
    status: str
    repository: str
    branch_name: str
    pr_url: str | None
    summary: str | None
    created_at: str


@dataclass
class Message:
    id: str
    role: str
    text: str


def _compute_delay(attempt: int) -> float:
    delay = min(INITIAL_DELAY * (BACKOFF_MULTIPLIER ** attempt), MAX_DELAY)
    jitter = delay * JITTER_FACTOR
    return delay + random.uniform(-jitter, jitter)


def _compute_429_delay(attempt: int, resp: requests.Response) -> float:
    delay = _compute_delay(attempt)
    retry_after = resp.headers.get("Retry-After")
    if retry_after is not None:
        delay = max(delay, float(retry_after))
    return delay


class CursorClient:
    def __init__(self, api_key: str):
        self._session = requests.Session()
        self._session.auth = (api_key, "")

    def _request(self, method: str, path: str, **kwargs: object) -> requests.Response:
        url = f"{BASE_URL}{path}"
        retries_429 = 0
        retries_transient = 0

        while True:
            resp = self._session.request(method, url, **kwargs)

            if resp.status_code == 429 and retries_429 < MAX_RETRIES_429:
                delay = _compute_429_delay(retries_429, resp)
                retries_429 += 1
                logger.warning(f"Rate limited (429). Retry {retries_429}/{MAX_RETRIES_429} in {delay:.1f}s")
                time.sleep(delay)
                continue

            if resp.status_code == 429:
                raise RateLimitError("Cursor API rate limit exceeded after max retries")

            if resp.status_code in TRANSIENT_CODES and retries_transient < MAX_RETRIES_TRANSIENT:
                delay = _compute_delay(retries_transient)
                retries_transient += 1
                logger.warning(f"Transient error ({resp.status_code}). Retry {retries_transient}/{MAX_RETRIES_TRANSIENT} in {delay:.1f}s")
                time.sleep(delay)
                continue

            if resp.status_code in TRANSIENT_CODES:
                raise AgentAPIError(f"Transient error {resp.status_code} after max retries", resp.status_code)

            if resp.status_code == 404:
                raise AgentNotFound()

            if resp.status_code >= 400:
                raise AgentAPIError(f"Cursor API error: {resp.status_code} {resp.text[:500]}", resp.status_code)

            return resp

    def launch_agent(
        self,
        prompt: str,
        repository: str,
        ref: str,
        model: str,
        branch_name: str,
        auto_pr: bool,
    ) -> AgentInfo:
        body = {
            "prompt": {"text": prompt},
            "model": model,
            "source": {"repository": repository, "ref": ref},
            "target": {
                "autoCreatePr": auto_pr,
                "branchName": branch_name,
                "openAsCursorGithubApp": True,
                "skipReviewerRequest": True,
            },
        }
        resp = self._request("POST", "/v0/agents", json=body)
        return _parse_agent_info(resp.json())

    def get_agent(self, agent_id: str) -> AgentInfo:
        resp = self._request("GET", f"/v0/agents/{agent_id}")
        return _parse_agent_info(resp.json())

    def list_agents(self, limit: int = 100) -> list[AgentInfo]:
        resp = self._request("GET", "/v0/agents", params={"limit": limit})
        data = resp.json()
        return [_parse_agent_info(a) for a in data.get("agents", [])]

    def get_conversation(self, agent_id: str) -> list[Message]:
        resp = self._request("GET", f"/v0/agents/{agent_id}/conversation")
        data = resp.json()
        return [
            Message(
                id=m.get("id", ""),
                role="user" if m.get("type") == "user_message" else "assistant",
                text=m.get("text", ""),
            )
            for m in data.get("messages", [])
        ]

    def send_followup(self, agent_id: str, prompt: str) -> None:
        self._request("POST", f"/v0/agents/{agent_id}/followup", json={"prompt": {"text": prompt}})

    def stop_agent(self, agent_id: str) -> None:
        self._request("POST", f"/v0/agents/{agent_id}/stop")

    def delete_agent(self, agent_id: str) -> None:
        self._request("DELETE", f"/v0/agents/{agent_id}")

    def list_models(self) -> list[str]:
        resp = self._request("GET", "/v0/models")
        return resp.json()


def _parse_agent_info(data: dict) -> AgentInfo:
    target = data.get("target", {})
    source = data.get("source", {})
    return AgentInfo(
        id=data.get("id", ""),
        name=data.get("name", ""),
        status=data.get("status", "UNKNOWN"),
        repository=source.get("repository", ""),
        branch_name=target.get("branchName", ""),
        pr_url=target.get("prUrl"),
        summary=data.get("summary"),
        created_at=data.get("createdAt", ""),
    )

File: src/api/cursor-client.ts
import { setTimeout as delay } from "node:timers/promises";

const BASE_URL = "https://api.cursor.com";

const INITIAL_DELAY = 5.0;
const BACKOFF_MULTIPLIER = 2.0;
const MAX_DELAY = 60.0;
const MAX_RETRIES_429 = 5;
const MAX_RETRIES_TRANSIENT = 3;
const JITTER_FACTOR = 0.2;
const TRANSIENT_CODES = new Set([502, 503, 504]);

export class AgentAPIError extends Error {
  statusCode: number;
  constructor(message: string, statusCode = 0) {
    super(message);
    this.name = "AgentAPIError";
    this.statusCode = statusCode;
  }
}

export class AgentNotFound extends AgentAPIError {
  constructor(agentId = "") {
    super(`Agent not found: ${agentId}`, 404);
    this.name = "AgentNotFound";
  }
}

export class RateLimitError extends AgentAPIError {
  retryAfter: number;
  constructor(message = "Rate limit exceeded", retryAfter = 0) {
    super(message, 429);
    this.name = "RateLimitError";
    this.retryAfter = retryAfter;
  }
}

export interface AgentInfo {
  id: string;
  name: string;
  status: string;
  repository: string;
  branch_name: string;
  pr_url: string | null;
  summary: string | null;
  created_at: string;
}

export interface Message {
  id: string;
  role: string;
  text: string;
}

function computeDelay(attempt: number): number {
  const d = Math.min(INITIAL_DELAY * BACKOFF_MULTIPLIER ** attempt, MAX_DELAY);
  const jitter = d * JITTER_FACTOR;
  return d + (Math.random() * 2 - 1) * jitter;
}

function compute429Delay(attempt: number, retryAfterHeader: string | null): number {
  let d = computeDelay(attempt);
  if (retryAfterHeader !== null) {
    const v = parseFloat(retryAfterHeader);
    if (!Number.isNaN(v)) {
      d = Math.max(d, v);
    }
  }
  return d;
}

function parseAgentInfo(data: Record<string, unknown>): AgentInfo {
  const target = (data.target as Record<string, unknown>) || {};
  const source = (data.source as Record<string, unknown>) || {};
  return {
    id: (data.id as string) ?? "",
    name: (data.name as string) ?? "",
    status: (data.status as string) ?? "UNKNOWN",
    repository: (source.repository as string) ?? "",
    branch_name: (target.branchName as string) ?? "",
    pr_url: (target.prUrl as string) ?? null,
    summary: (data.summary as string) ?? null,
    created_at: (data.createdAt as string) ?? "",
  };
}

export class CursorClient {
  private readonly apiKey: string;

  constructor(apiKey: string) {
    this.apiKey = apiKey;
  }

  private async request(method: string, path: string, body?: unknown): Promise<Response> {
    const url = `${BASE_URL}${path}`;
    let retries429 = 0;
    let retriesTransient = 0;
    const headers: Record<string, string> = {
      Authorization: `Basic ${Buffer.from(`${this.apiKey}:`).toString("base64")}`,
      "Content-Type": "application/json",
    };
    while (true) {
      const resp = await fetch(url, {
        method,
        headers,
        body: body !== undefined ? JSON.stringify(body) : undefined,
      });
      if (resp.status === 429 && retries429 < MAX_RETRIES_429) {
        const d = compute429Delay(retries429, resp.headers.get("Retry-After"));
        retries429 += 1;
        await delay(Math.ceil(d * 1000));
        continue;
      }
      if (resp.status === 429) {
        throw new RateLimitError("Cursor API rate limit exceeded after max retries");
      }
      if (TRANSIENT_CODES.has(resp.status) && retriesTransient < MAX_RETRIES_TRANSIENT) {
        const d = computeDelay(retriesTransient);
        retriesTransient += 1;
        await delay(Math.ceil(d * 1000));
        continue;
      }
      if (TRANSIENT_CODES.has(resp.status)) {
        throw new AgentAPIError(`Transient error ${resp.status} after max retries`, resp.status);
      }
      if (resp.status === 404) {
        throw new AgentNotFound();
      }
      if (resp.status >= 400) {
        const text = await resp.text();
        throw new AgentAPIError(`Cursor API error: ${resp.status} ${text.slice(0, 500)}`, resp.status);
      }
      return resp;
    }
  }

  async launchAgent(
    prompt: string,
    repository: string,
    ref: string,
    model: string,
    branchName: string,
    autoPr: boolean,
  ): Promise<AgentInfo> {
    const body = {
      prompt: { text: prompt },
      model,
      source: { repository, ref },
      target: {
        autoCreatePr: autoPr,
        branchName,
        openAsCursorGithubApp: true,
        skipReviewerRequest: true,
      },
    };
    const resp = await this.request("POST", "/v0/agents", body);
    const data = (await resp.json()) as Record<string, unknown>;
    return parseAgentInfo(data);
  }

  async getAgent(agentId: string): Promise<AgentInfo> {
    const resp = await this.request("GET", `/v0/agents/${agentId}`);
    const data = (await resp.json()) as Record<string, unknown>;
    return parseAgentInfo(data);
  }

  async listAgents(limit = 100): Promise<AgentInfo[]> {
    const resp = await this.request("GET", `/v0/agents?limit=${limit}`);
    const data = (await resp.json()) as { agents?: Record<string, unknown>[] };
    return (data.agents ?? []).map((a) => parseAgentInfo(a));
  }

  async getConversation(agentId: string): Promise<Message[]> {
    const resp = await this.request("GET", `/v0/agents/${agentId}/conversation`);
    const data = (await resp.json()) as { messages?: Record<string, unknown>[] };
    return (data.messages ?? []).map((m) => ({
      id: (m.id as string) ?? "",
      role: m.type === "user_message" ? "user" : "assistant",
      text: (m.text as string) ?? "",
    }));
  }

  async sendFollowup(agentId: string, prompt: string): Promise<void> {
    await this.request("POST", `/v0/agents/${agentId}/followup`, { prompt: { text: prompt } });
  }

  async stopAgent(agentId: string): Promise<void> {
    await this.request("POST", `/v0/agents/${agentId}/stop`);
  }

  async deleteAgent(agentId: string): Promise<void> {
    await this.request("DELETE", `/v0/agents/${agentId}`);
  }

  async listModels(): Promise<unknown> {
    const resp = await this.request("GET", "/v0/models");
    return resp.json();
  }
}
